sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -50,15 +51,23 @@ import org.apache.spark.unsafe.types.{ByteArray, UTF8String}
   """)
 case class Concat(children: Seq[Expression]) extends Expression with ImplicitCastInputTypes {
 
-  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.size)(StringType)
-  override def dataType: DataType = StringType
+  private lazy val isBinaryMode = children.nonEmpty && children.forall(_.dataType == BinaryType)
+
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq.fill(children.size)(if (isBinaryMode) BinaryType else StringType)
+  override def dataType: DataType = if (isBinaryMode) BinaryType else StringType
```

Comment:
Shouldn't we worry about backward compatibility?

Reply:
yea, should be.
Any existing option for keeping back compatibility? One option I think is to use a new hive type coercion option? https://issues.apache.org/jira/browse/SPARK-22722
Or, how about adding a new option?

Reply:
Conf is needed for sure. We also need a `Migration Guide`.

Reply:
ok

Reply:
Is it ok to add a new option for this case only? If we keep adding new options for each case, options could blow up?